### PR TITLE
Get preload blocks when its needed

### LINKED ISSUE #2802: Too many memory usage problem when preloading
# Context
In PBFT testing, some validators have problems when following the highest tip with preload.
Additionally, the one validators have an OS handle issue when exiting the node.
At the request of @longfin, I came across an OS handle issue.

And In my opinion, what may cause an OS handling issue is a non-exists block.
Because even preload is finished, cannot find blocks from `BlockChain<T>`.

So I dig preload phase, and I found a memory usage problem.

# Problem
In `Swarm<T>.CompleteBlockAsync()`, when called the `Preload` step, It uses too much memory when getting `deltaBlocks` from `IStore` https://github.com/planetarium/libplanet/blob/ff823f8435133689fcbd3f26508e40a2eab969ae/Libplanet.Net/Swarm.BlockSync.cs#L480-L492

I suggest using `List<BlockHash>` instead of `LinkedList<Block<T>>` for uses less memory.

### PR DESCRIPTION
**Pull request by riemannulus**
relative #2802

When finding preload branchpoint, find only `BlockHash` instead of `Block<T>`

### CODE REVIEW

File: Libplanet.Net/Swarm.BlockSync.cs [hunk 1]
```diff
@@ -476,26 +477,47 @@ in completedBlocks.WithCancellation(cancellationToken))
...
 
+                if (!(tipCandidate is { }))
+                {
+                    throw new SwarmException(
+                        $"Tip candidate is null.");
+                }
+
                 // NOTE: All blocks in deltaBlocks will have non-null previousHash.
-                var deltaBlocks = new LinkedList<Block<T>>(new[] { tipCandidate });
+                BlockChainSlice blockChainSlice = new BlockChainSlice(new[] { tipCandidate.Hash });
```
Comment (dahlia) [inline Libplanet.Net/Swarm.BlockSync.cs:487]: For conciseness:

```suggestion
                BlockChainSlice blockChainSlice = new (new[] { tipCandidate.Hash });
```

Or:


```suggestion
                var blockChainSlice = new BlockChainSlice(new[] { tipCandidate.Hash });
```

**Comment by riemannulus** [comment]
PTAL 🙏  @dahlia @greymistcube @longfin

**Review by dahlia — DISMISSED** [review]

File: Libplanet.Net/Swarm.BlockSync.cs [hunk 2]
```diff
@@ -476,26 +477,47 @@ in completedBlocks.WithCancellation(cancellationToken))
...
 
-                cancellationToken.ThrowIfCancellationRequested();
+                blockChainSlice.Reverse();
+                Block<T> firstBlock = workspace.Store.GetBlock<T>(blockChainSlice.First());
+
+                if (!(firstBlock.PreviousHash is { }))
+                {
+                    throw new SwarmException(
+                        "A genesis block cannot be a preload target block.");
+                }
```
Comment (longfin) [inline Libplanet.Net/Swarm.BlockSync.cs:516]:
```suggestion
                blockChainSlice.Reverse();
                BlockDigest firstBlock = workspace.Store.GetBlockDigest(blockChainSlice.First());

                if (!(firstBlock?.PreviousHash is { } previousBlockHash))
                {
                    throw new SwarmException(
                        "A genesis block cannot be a preload target block.");
                }
```

A trivial fix.

File: Libplanet.Net/Swarm.BlockSync.cs [hunk 3]
```diff
@@ -476,26 +477,47 @@ in completedBlocks.WithCancellation(cancellationToken))
...
 
-                branchpoint = workspace[deltaBlocks.First().PreviousHash.Value];
+                branchpoint = workspace[firstBlock.PreviousHash.Value];
```
Comment (longfin) [inline Libplanet.Net/Swarm.BlockSync.cs:518]:
```suggestion
                branchpoint = workspace[previousBlockHash];
```

File: Libplanet.Net/Swarm.BlockSync.cs [hunk 4]
```diff
@@ -476,26 +477,47 @@ in completedBlocks.WithCancellation(cancellationToken))
...
                 while (true)
                 {
-                    Block<T> b = deltaBlocks.First();
+                    BlockHash bh = blockChainSlice.Last();
+                    if (!workspace.Store.ContainsBlock(bh))
+                    {
+                        break;
+                    }
+
+                    Block<T> b = workspace.Store.GetBlock<T>(bh);
```
Comment (longfin) [inline Libplanet.Net/Swarm.BlockSync.cs:496]:
```suggestion
                    BlockDigest? b = workspace.Store.GetBlockDigest(bh);
```

File: Libplanet.Net/Swarm.BlockSync.cs [hunk 5]
```diff
@@ -476,26 +477,47 @@ in completedBlocks.WithCancellation(cancellationToken))
...
-                var deltaBlocks = new LinkedList<Block<T>>(new[] { tipCandidate });
+                var blockChainSlice = new BlockChainSlice(new[] { tipCandidate.Hash });
                 while (true)
                 {
-                    Block<T> b = deltaBlocks.First();
+                    BlockHash bh = blockChainSlice.Last();
+                    if (!workspace.Store.ContainsBlock(bh))
+                    {
+                        break;
+                    }
```
Comment (longfin) [inline Libplanet.Net/Swarm.BlockSync.cs:494]: Is this check needed? I think that we're checking this with `.GetBlock<T>()` or `GetBlockDigest()`.

File: Libplanet.Net/Swarm.BlockSync.cs [hunk 6]
```diff
@@ -476,26 +477,47 @@ in completedBlocks.WithCancellation(cancellationToken))
...
                     else
                     {
                         break;
                     }
+
+                    cancellationToken.ThrowIfCancellationRequested();
                 }
 
-                cancellationToken.ThrowIfCancellationRequested();
+                blockChainSlice.Reverse();
```
Comment (longfin) [inline Libplanet.Net/Swarm.BlockSync.cs:509]: If we construct `blockChainSlice` from `.First()` (as previously), this `.Reverse()` doesn't seem to need. is it right what I think?

File: Libplanet.Net/Swarm.BlockSync.cs (hunk 5 above)
Comment (riemannulus) [reply Libplanet.Net/Swarm.BlockSync.cs:494]: I forgot to remove this check 😓 I'll remove this check next commit.

File: Libplanet.Net/Swarm.BlockSync.cs (hunk 6 above)
Comment (riemannulus) [reply Libplanet.Net/Swarm.BlockSync.cs:509]: Previously, the data structure for memorizing target block is `LinkedList<T>` but in this PR, changed to `List<T>`. So I am worried about the performance issue when appending each item to `List<T>`'s index 0.

May we use `LinkedList<T>` again?

Comment (riemannulus) [reply Libplanet.Net/Swarm.BlockSync.cs:509]: I've been thinking about it, and changing it to `LinkedList<T>` looks good. I'll roll back to `LinkedList<T>`.

Comment (riemannulus) [reply Libplanet.Net/Swarm.BlockSync.cs]: ec496acdd1366ea75a3fed132e352d5f47354809 amended

File: Libplanet.Net/Swarm.BlockSync.cs (hunk 5 above)
Comment (riemannulus) [reply Libplanet.Net/Swarm.BlockSync.cs]: ec496acdd1366ea75a3fed132e352d5f47354809 amended

**Review by longfin — APPROVED** [review]

**Pull request merged**